### PR TITLE
[Core]Extending  BinbasedFastPointLocator to any geometry using inner methods fo the geometries

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

**Comment by loumalouomega** [comment]
@bodhinandach https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ParticleMechanicsApplication/custom_utilities/quad_binbased_fast_point_locator.h can be removed, but maybe there is still necessary to improve the PointLocalCoordinates form the Quad2D

**Comment by loumalouomega** [comment]
Here is the implementation of the Quad2D, maybe we need to specialize:

https://www.colorado.edu/engineering/CAS/courses.d/IFEM.d/IFEM.Ch17.d/IFEM.Ch17.pdf

File: kratos/utilities/binbased_fast_point_locator.h [hunk 1]
```diff
@@ -203,132 +208,80 @@ class BinBasedFastPointLocator
...
+            // Loop over the candidate elements and check if the particle falls within
+            for (IndexType i = 0; i < results_found; i++) {
+                GeometryType& geom = (*(ItResultBegin + i))->GetGeometry();
 
-        N[0] = CalculateVol(x1, y1, x2, y2, xc, yc) * inv_area;
-        N[1] = CalculateVol(x2, y2, x0, y0, xc, yc) * inv_area;
-        N[2] = CalculateVol(x0, y0, x1, y1, xc, yc) * inv_area;
+                // Find local position
+                array_1d<double, 3> point_local_coordinates;
+                const bool is_found = geom.IsInside(rCoordinates, point_local_coordinates, Tolerance);
```
Comment (bodhinandach) [inline kratos/utilities/binbased_fast_point_locator.h]: Thanks @loumalouomega So you are generalizing the point locator by using the IsInside function right... I will try to see whether I can implement the PointLocalCoordinates for Quad2D. But thanks for the reference, will take a look at it.

**Review by bodhinandach — APPROVED** [review]

**Review by roigcarlo — APPROVED** [review]
:+1: Thx!

File: kratos/utilities/convect_particles_utilities.h [hunk 2]
```diff
@@ -112,7 +113,9 @@ template<std::size_t TDim> class ParticleConvectUtily
 
                 if(substep > 1 ) //first check if it falls within the same element
                 {
-                    is_found = mpSearchStructure->CalculatePosition(pelement->GetGeometry(), current_position[0], current_position[1], current_position[2], N);
+                    Geometry< Node < 3 > >& geom = pelement->GetGeometry();
+                    is_found = geom.IsInside(current_position, aux_point_local_coordinates, 1.0e-5);
```
Comment (roigcarlo) [inline kratos/utilities/convect_particles_utilities.h:117]: Actually, I have one comment:
Shouldn't this be a double::epsilon or whatever, 1.0e-5 seems a little bit magic

Comment (loumalouomega) [reply kratos/utilities/convect_particles_utilities.h:117]: This was the value already existing in the BinbasedPointLocator, so the best way to not break code it is to use the same value

File: kratos/utilities/binbased_fast_point_locator.h [hunk 3]
```diff
@@ -122,225 +161,207 @@ class BinBasedFastPointLocator
...
-    ///and return a pointer to the element and the vector containing the
-    ///shape functions that define the postion within the element
-    ///if "false" is devolved the element is not found
-    ///REMARK: this function is threadsafe and can be used within OpenMP loops
-    bool FindPointOnMesh(const array_1d<double, 3 >& coords,
-                         array_1d<double, TDim + 1 > & N,
-                         Element::Pointer& pelement,
-                         ResultIteratorType result_begin,
-                         const unsigned int MaxNumberOfResults,
-                         const double Tolerance = 1.0e-5
```
Comment (loumalouomega) [inline kratos/utilities/binbased_fast_point_locator.h:135]: @roigcarlo The original magic number is here

**Pull request merged**